### PR TITLE
CET-13421 Adobe performance with Backstage on Bird's Eye report

### PR DESCRIPTION
**Pull request by voytekgar**
## Change description

CET-13421 Adobe performance with Backstage on Bird's Eye report

**Requires BE**: https://github.com/cortexapps/brain-backend/pull/8087

Changed fetching owners for entities to fetch only those for entities for which we have scores for scorecard.

## Type of change

- [ ] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)
- [x] Improvement (improves codebase without changing functionality)

### Development

- [x] The changelog has been updated as appropriate
- [x] Lint rules pass locally
- [x] Application changes have been tested thoroughly
- [x] Automated tests covering modified code pass

### Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [x] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [x] Pull request linked to task tracker where applicable

### CODE REVIEW

File: src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx [hunk 1]
```diff
@@ -113,7 +109,13 @@ export const HeatmapPage = () => {
     api.getAllTeams(),
   );
   const { value: teamsByEntityId, loading: isLoadingTeamsByEntityId } =
-    useCortexApi(api => api.getAllTeamsByEntityId());
+    useCortexApiMemorized(
+      async api =>
+        scores && scores.length > 0
```
Comment (igorRoog) [inline src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:114]:
```suggestion
        !isNil(scores) && !isEmpty(scores)
```

File: src/utils/hooks.ts [hunk 2]
```diff
@@ -244,6 +220,37 @@ export function useCortexApi<T>(
...
+  deps: any[] = [],
+) {
+  const cortexApi = useApi(cortexApiRef);
+
+  return useAsyncMemorized(async () => {
+    return await f(cortexApi);
+  }, deps);
+}
+
+function useAsyncMemorized<T>(f: () => Promise<T>, deps: any[]) {
```
Comment (igorRoog) [inline src/utils/hooks.ts:234]:
```suggestion
function useAsyncMemoized<T>(f: () => Promise<T>, deps: any[]) {
```

File: src/utils/hooks.ts [hunk 3]
```diff
@@ -244,6 +220,37 @@ export function useCortexApi<T>(
   }, deps);
 }
 
+export function useCortexApiMemorized<T>(
```
Comment (igorRoog) [inline src/utils/hooks.ts:223]:
```suggestion
export function useCortexApiMemoized<T>(
```

File: src/utils/hooks.ts [hunk 4]
```diff
@@ -244,6 +220,37 @@ export function useCortexApi<T>(
...
+  }, deps);
+}
+
+function useAsyncMemorized<T>(f: () => Promise<T>, deps: any[]) {
+  const [cache, setCache] = useState<Record<string, any>>({});
+  const cacheKey = useMemo(() => JSON.stringify(deps), [deps]);
+  const cachedResult = useMemo(() => cache[cacheKey], [cacheKey, cache]);
+
+  return useAsync(async () => {
+    if (cachedResult !== undefined) {
```
Comment (igorRoog) [inline src/utils/hooks.ts:240]:
```suggestion
    if (!isNil(cachedResult)) {
```

File: src/utils/hooks.ts [hunk 5]
```diff
@@ -244,6 +220,37 @@ export function useCortexApi<T>(
...
+
+  return useAsyncMemorized(async () => {
+    return await f(cortexApi);
+  }, deps);
+}
+
+function useAsyncMemorized<T>(f: () => Promise<T>, deps: any[]) {
+  const [cache, setCache] = useState<Record<string, any>>({});
+  const cacheKey = useMemo(() => JSON.stringify(deps), [deps]);
+  const cachedResult = useMemo(() => cache[cacheKey], [cacheKey, cache]);
```
Comment (igorRoog) [inline src/utils/hooks.ts:237]: It doesn't look very complex to access an array, so I guess it doesn't have to be memoized, right?

**Review by igorRoog — COMMENTED** [review]
Couple of nits.

**Comment by browniefed** [comment]
One thing we actually want to do is swap this over to `/scores/compact` which sends a way smaller data structure that is reconstructed on the frontend.

That may be worth doing opposed to this.

**Comment by voytekgar** [comment]
@browniefed 

> One thing we actually want to do is swap this over to /scores/compact which sends a way smaller data structure that is reconstructed on the frontend.
> That may be worth doing opposed to this.

Actually the main problem is not the amount of entities to fetch the owners for but resolving ownerships on BE. This will be resolved with the graph for long-term.

**Review by igorRoog — APPROVED** [review]

**Pull request merged**